src/c++/perf_analyzer/client_backend/openai/openai_client.cc

```diff
@@ -0,0 +1,297 @@
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// Include this first to make sure we are a friend of common classes.
+#define TRITON_INFERENCE_SERVER_CLIENT_CLASS InferenceServerHttpClient
+#include "openai_client.h"
+
+#include <curl/curl.h>
+
+#include <algorithm>
+#include <atomic>
+#include <cctype>
+#include <climits>
+#include <cstdint>
+#include <deque>
+#include <iostream>
+#include <string>
+#include <utility>
+
+#include "common.h"
+
+#ifdef TRITON_ENABLE_ZLIB
+#include <zlib.h>
+#endif
+
+extern "C" {
+#include "cencode.h"
+}
+
+#ifdef _WIN32
+#define strncasecmp(x, y, z) _strnicmp(x, y, z)
+#undef min  // NOMINMAX did not resolve std::min compile error
+#endif      //_WIN32
+
+namespace triton { namespace perfanalyzer { namespace clientbackend {
+namespace openai {
+
+//==============================================================================
+
+void
+ChatCompletionRequest::SendResponse(bool is_final, bool is_null)
+{
+  response_callback_(new ChatCompletionResult(
+      http_code_, std::move(response_buffer_), is_final, is_null, request_id_));
+}
+
+ChatCompletionClient::ChatCompletionClient(
+    const std::string& url, const std::string& endpoint, bool verbose,
+    const HttpSslOptions& ssl_options)
+    : HttpClient(std::string(url + "/" + endpoint), verbose, ssl_options)
+{
+}
+
+size_t
+ChatCompletionClient::RequestProvider(
+    void* contents, size_t size, size_t nmemb, void* userp)
+{
+  auto request = reinterpret_cast<ChatCompletionRequest*>(userp);
+
+  size_t input_bytes = 0;
+  request->GetNextInput(
+      reinterpret_cast<uint8_t*>(contents), size * nmemb, &input_bytes);
+
+  if (input_bytes == 0) {
+    request->timer_.CaptureTimestamp(
+        triton::client::RequestTimers::Kind::SEND_END);
+  }
+
+  return input_bytes;
+}
+
+size_t
+ChatCompletionClient::ResponseHeaderHandler(
+    void* contents, size_t size, size_t nmemb, void* userp)
+{
+  auto request = reinterpret_cast<ChatCompletionRequest*>(userp);
+
+  char* buf = reinterpret_cast<char*>(contents);
+  size_t byte_size = size * nmemb;
+
+  std::string hdr(buf, byte_size);
+  std::transform(hdr.begin(), hdr.end(), hdr.begin(), [](unsigned char c) {
+    return std::tolower(c);
+  });
+  if (hdr.find("content-type") != std::string::npos) {
+    request->is_stream_ = (hdr.find("text/event-stream") != std::string::npos);
+  }
+
+  return byte_size;
+}
+
+size_t
+ChatCompletionClient::ResponseHandler(
+    void* contents, size_t size, size_t nmemb, void* userp)
+{
+  // [WIP] verify if the SSE responses received are complete, or the response
```

Comment:
Will this be included in this ticket? If so, we should block the merge on this to avoid forgetting to finish this prior to merging.

Reply:
The assumption holds for vllm server. It's kind of hand waving but I would want to skip the stitching until we find the case that the response are chunked in the middle, as stitching requires examination on each response chunk (overhead).

Reply:
Would this be a future ticket then? I feel like we shouldn't be merging with a [WIP] unless this links to an open ticket for future work.

Reply:
Good catch DY. I looked for TODO and FIXME but missed this WIP. 
@GuanLuo is there more testing or validation to do here? When we bring up a NIM server should we look/test for anything specific? Is there a follow-on ticket to file?

Ryan O in a meeting yesterday mentioned some things about some implementations possibly sending empty packets that should be ignored (including possibly immediately after receiving the response), as well as possibly sending keep-alive signals that should be ignored.

Let me know what you think the risks are for MVP and what we need to test and implement later.

Reply:
I would print out the received responses from `SendResponse()` to make sure NIM doesn't chunk the HTTP responses in the way that misaligns with the SSE responses. The same applies to connection to other OpenAI server for the MVP.

For empty message, we can improve the write callback now that to ignore the invocation if calcualted `size_t result_bytes = size * nmemb;` is 0.

Future item to make the write callback bulletproof is to assume the response can be chunked at arbitrary position, then bake in checking for SSE style (`data: .*\n\n`) by iterating all received buffer character by character.

Reply:
The comment was updated, and it now points to TMA-1666